pkg/pprof/cell.go

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package pprof
+
+import (
+	"errors"
+	"net"
+	"net/http"
+	_ "net/http/pprof"
+	"strconv"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+)
+
+const (
+	// Pprof is the flag to enable the registration of pprof HTTP handlers
+	Pprof = "pprof"
+
+	// PprofAddress is the flag to set the address that pprof listens on
+	PprofAddress = "pprof-address"
+
+	// PprofPort is the flag to set the port that pprof listens on
+	PprofPort = "pprof-port"
+)
+
+type Server interface {
+	// Port returns the port at which the server is listening
+	Port() int
+}
+
+// Cell creates the cell for pprof, that registers its HTTP handlers to serve
+// profiling data in the format expected by the pprof visualization tool.
+var Cell = cell.Module(
+	"pprof",
+	"pprof HTTP handlers",
+
+	// We don't call cell.Config directly here, because the operator
+	// uses different flags names for the same pprof config flags.
+	// Therefore, to register each flag with the correct name, we leave
+	// the call to cell.Config to the user of the cell.
+
+	// Provide coupled with Invoke is used to improve cell testability,
+	// namely to allow taking a reference to the Server and call Port() on it.
+	cell.Provide(newServer),
+	cell.Invoke(func(srv Server) {}),
+)
+
+// Config contains the configuration for the pprof cell.
+type Config struct {
+	Pprof        bool
+	PprofAddress string
+	PprofPort    int
+}
+
+func (def Config) Flags(flags *pflag.FlagSet) {
+	flags.Bool(Pprof, def.Pprof, "Enable serving pprof debugging API")
+	flags.String(PprofAddress, def.PprofAddress, "Address that pprof listens on")
+	flags.Int(PprofPort, def.PprofPort, "Port that pprof listens on")
+}
+
+func newServer(lc hive.Lifecycle, log logrus.FieldLogger, cfg Config) Server {
+	if !cfg.Pprof {
+		return nil
+	}
+
+	srv := &server{
+		logger:  log,
+		address: cfg.PprofAddress,
+		port:    cfg.PprofPort,
+	}
+	lc.Append(srv)
+
+	return srv
+}
+
+type server struct {
+	logger logrus.FieldLogger
+
+	address string
+	port    int
+
+	httpSrv  *http.Server
+	listener net.Listener
+}
+
+func (s *server) Start(ctx hive.HookContext) error {
+	listener, err := net.Listen("tcp", net.JoinHostPort(s.address, strconv.Itoa(s.port)))
+	if err != nil {
+		return err
+	}
+	s.listener = listener
+
+	log = log.WithFields(logrus.Fields{
+		"ip":   s.listener.Addr().(*net.TCPAddr).IP,
+		"port": s.listener.Addr().(*net.TCPAddr).Port,
+	})
+
+	s.httpSrv = &http.Server{}
```

Comment:
I'm slightly worried about the use of the default mux here. Consider instead building a new mux here with just the handlers for pprof instead so we'll know exactly what we're serving and not potentially random stuff added by imported packages.

Check pprof.init for what the handlers need to be: https://cs.opensource.google/go/go/+/refs/tags/go1.20.1:src/net/http/pprof/pprof.go;l=85;bpv=1